### DISCUSSION TITLE
Websocket no upgrade header present

### DISCUSSION THREAD

**Discussion by SamTV12345** [actix/actix-web#2998] [Q&A]
Following up  [this](https://github.com/actix/actix-web/discussions/2691) question. How can I add headers to ws::start that work with only the actix web server? I don't want to add an nginx proxy in front of the actix web server.

I followed [this](https://github.com/antholeole/actix-sockets/tree/main/src) tutorial but when I call the url it states `No WebSocket Upgrade header found`.

**Comment by Samielakkad** [comment]
`No WebSocket Upgrade header found` usually means the route was reached by a normal HTTP request, not by a WebSocket handshake.

`ws::start(...)` does not add upgrade headers to an ordinary request. It checks that the incoming request already contains the WebSocket upgrade headers:

```text
Connection: Upgrade
Upgrade: websocket
Sec-WebSocket-Key: ...
Sec-WebSocket-Version: 13
```

Those headers are normally sent by a WebSocket client, not by the server.

So this will fail:

```text
curl http://localhost:8080/ws
```

or opening the URL directly in a browser address bar, because both are plain HTTP GET requests.

Use a WebSocket client instead, for example browser JavaScript:

```js
const socket = new WebSocket("ws://127.0.0.1:8080/ws");

socket.onopen = () => {
  console.log("connected");
  socket.send("hello");
};

socket.onmessage = (event) => {
  console.log("message", event.data);
};
```

Or a CLI WebSocket client such as `websocat`:

```bash
websocat ws://127.0.0.1:8080/ws
```

A minimal Actix route looks like:

```rust
async fn ws_index(
    req: HttpRequest,
    stream: web::Payload,
) -> Result<HttpResponse, Error> {
    ws::start(MyWsActor, &req, stream)
}

App::new().route("/ws", web::get().to(ws_index))
```

If you are behind a proxy, then the proxy must preserve WebSocket upgrade headers. But if you are connecting directly to the Actix server, no nginx or extra proxy is required. The important part is simply: connect with `ws://` or `wss://` using an actual WebSocket client, not with a normal HTTP request.
